From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Henry Lin <henryl@nvidia.com>
Date: Fri, 3 May 2024 17:08:06 +0000
Subject: [PATCH 089/245] NVIDIA: SAUCE: xhci: tegra: enable firmware log

enable XUSB fwlog

http://nvbugs/4613856

Change-Id: I403d96ed47d970a3d75cca18b624aeb752744698
Signed-off-by: Henry Lin <henryl@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/linux-stable/+/3131824
(cherry picked from commit a18be1a903d3ede67977101bb74e573213131301)
Signed-off-by: Bodla Rakesh Babu <rbodla@nvidia.com>
Reviewed-on: https://git-master.nvidia.com/r/c/3rdparty/canonical/linux-noble/+/3178748
GVS: buildbot_gerritrpt <buildbot_gerritrpt@nvidia.com>
Reviewed-by: Laxman Dewangan <ldewangan@nvidia.com>
(cherry picked from commit 6ae34c8204121f69e7d3fe86503405bfd58cb9fa)

diff --git a/drivers/usb/host/xhci-tegra.c b/drivers/usb/host/xhci-tegra.c
index eb1f2cc75ebf2..a8474d2825651 100644
--- a/drivers/usb/host/xhci-tegra.c
+++ b/drivers/usb/host/xhci-tegra.c
@@ -2,12 +2,14 @@
 /*
  * NVIDIA Tegra xHCI host controller driver
  *
- * Copyright (c) 2014-2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ * SPDX-FileCopyrightText: Copyright (c) 2014-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
  * Copyright (C) 2014 Google, Inc.
  */
 
 #include <linux/clk.h>
+#include <linux/circ_buf.h>
 #include <linux/delay.h>
+#include <linux/debugfs.h>
 #include <linux/dma-mapping.h>
 #include <linux/firmware.h>
 #include <linux/interrupt.h>
@@ -30,6 +32,7 @@
 #include <linux/usb/otg.h>
 #include <linux/usb/phy.h>
 #include <linux/usb/role.h>
+#include <linux/vmalloc.h>
 #include <soc/tegra/pmc.h>
 
 #include "xhci.h"
@@ -53,6 +56,7 @@
 #define XUSB_CFG_AXI_CFG			0x0f8
 #define XUSB_CFG_ARU_C11_CSBRANGE		0x41c
 #define XUSB_CFG_ARU_CONTEXT			0x43c
+#define XUSB_CFG_ARU_FW_SCRATCH                 0x440
 #define XUSB_CFG_ARU_CONTEXT_HS_PLS		0x478
 #define XUSB_CFG_ARU_CONTEXT_FS_PLS		0x47c
 #define XUSB_CFG_ARU_CONTEXT_HSFS_SPEED		0x480
@@ -131,6 +135,7 @@
 
 /* CSB ARU registers */
 #define XUSB_CSB_ARU_SCRATCH0			0x100100
+#define XUSB_CSB_ARU_SCRATCH1			0x100104
 
 /* MP CSB registers */
 #define XUSB_CSB_MP_ILOAD_ATTR			0x101a00
@@ -152,8 +157,31 @@
 
 #define IMEM_BLOCK_SIZE				256
 
-#define FW_IOCTL_TYPE_SHIFT			24
-#define FW_IOCTL_CFGTBL_READ		17
+#define FW_IOCTL_LOG_DEQUEUE_LOW        (4)
+#define FW_IOCTL_LOG_DEQUEUE_HIGH       (5)
+#define FW_IOCTL_CFGTBL_READ		(17)
+#define FW_IOCTL_INIT_LOG_BUF		(31)
+#define FW_IOCTL_LOG_DEQUEUE_IDX	(32)
+#define FW_IOCTL_TYPE_SHIFT             (24)
+#define FW_LOG_SIZE                     ((int) sizeof(struct log_entry))
+#define FW_LOG_COUNT                    (4096)
+#define FW_LOG_RING_SIZE                (FW_LOG_SIZE * FW_LOG_COUNT)
+#define FW_LOG_PAYLOAD_SIZE             (27)
+#define DRIVER                          (0x01)
+#define CIRC_BUF_SIZE                   (4 * (1 << 20)) /* 4MB */
+#define FW_LOG_THREAD_RELAX             (msecs_to_jiffies(500))
+
+/* tegra_xhci_firmware_log.flags bits */
+#define FW_LOG_CONTEXT_VALID            (0)
+#define FW_LOG_FILE_OPENED              (1)
+
+#define FW_MAJOR_VERSION(x)             (((x) >> 24) & 0xff)
+#define FW_MINOR_VERSION(x)             (((x) >> 16) & 0xff)
+
+enum build_info_log {
+	LOG_NONE = 0,
+	LOG_MEMORY
+};
 
 #define WAKE_IRQ_START_INDEX			2
 
@@ -188,7 +216,10 @@ struct tegra_xusb_fw_header {
 	u8 magic[8];
 	__le32 ss_low_power_entry_timeout;
 	u8 num_hsic_port;
-	u8 padding[139]; /* Pad to 256 bytes */
+	u8 ss_portmap;
+	u8 build_log:4;
+	u8 build_type:4;
+	u8 padding[137]; /* Pad to 256 bytes */
 };
 
 struct tegra_xusb_phy_type {
@@ -258,6 +289,28 @@ struct tegra_xusb_context {
 	u32 *fpci;
 };
 
+struct log_entry {
+	u32 sequence_no;
+	u8 data[FW_LOG_PAYLOAD_SIZE];
+	u8 owner;
+};
+
+struct tegra_xhci_firmware_log {
+	dma_addr_t phys_addr;           /* dma-able address */
+	void *virt_addr;                /* kernel va of the shared log buffer */
+	struct log_entry *dequeue;      /* current dequeue pointer (va) */
+	struct circ_buf circ;           /* big circular buffer */
+	u32 seq;                        /* log sequence number */
+
+	struct task_struct *thread;     /* a thread to consume log */
+	struct mutex mutex;
+	wait_queue_head_t read_wait;
+	wait_queue_head_t write_wait;
+	wait_queue_head_t intr_wait;
+	struct dentry *log_file;
+	unsigned long flags;
+};
+
 struct tegra_xusb {
 	struct device *dev;
 	void __iomem *regs;
@@ -317,6 +370,13 @@ struct tegra_xusb {
 		dma_addr_t phys;
 	} fw;
 
+	u8 build_log;
+	time64_t timestamp;
+	u32 version_id;
+
+	struct dentry *debugfs_dir;
+	struct tegra_xhci_firmware_log log;
+
 	bool suspended;
 	struct tegra_xusb_context context;
 	u8 lp0_utmi_pad_mask;
@@ -414,6 +474,434 @@ static void bar2_csb_writel(struct tegra_xusb *tegra, u32 value,
 	bar2_writel(tegra, value, XUSB_BAR2_CSB_BASE_ADDR + ofs);
 }
 
+/**
+ * fw_log_next - find next log entry in a tegra_xhci_firmware_log context.
+ *      This function takes care of wrapping. That means when current log entry
+ *      is the last one, it returns with the first one.
+ *
+ * @param log   The tegra_xhci_firmware_log context.
+ * @param this  The current log entry.
+ * @return      The log entry which is next to the current one.
+ */
+static inline struct log_entry *fw_log_next(
+		struct tegra_xhci_firmware_log *log, struct log_entry *this)
+{
+	struct log_entry *first = (struct log_entry *) log->virt_addr;
+	struct log_entry *last = first + FW_LOG_COUNT - 1;
+
+	WARN((this < first) || (this > last), "%s: invalid input\n", __func__);
+
+	return (this == last) ? first : (this + 1);
+}
+
+/**
+ * fw_log_update_dequeue_pointer - update dequeue pointer to both firmware and
+ *      tegra_xhci_firmware_log.dequeue.
+ *
+ * @param log   The tegra_xhci_firmware_log context.
+ * @param n     Counts of log entries to fast-forward.
+ */
+static inline void fw_log_update_deq_pointer(
+		struct tegra_xhci_firmware_log *log, int n)
+{
+	struct tegra_xusb *tegra =
+		container_of(log, struct tegra_xusb, log);
+	struct device *dev = tegra->dev;
+	struct log_entry *deq = tegra->log.dequeue;
+	dma_addr_t physical_addr;
+	u16 log_index;
+	u32 reg;
+
+	dev_vdbg(dev, "curr 0x%p fast-forward %d entries\n", deq, n);
+	while (n-- > 0)
+		deq = fw_log_next(log, deq);
+
+	tegra->log.dequeue = deq;
+	physical_addr = tegra->log.phys_addr +
+		((u8 *)deq - (u8 *)tegra->log.virt_addr);
+	log_index = (u16)((u8 *)deq - (u8 *)tegra->log.virt_addr) /
+			   sizeof(struct log_entry);
+
+	if (!tegra->soc->firmware) {
+		/* update 16 bit log index to firmware */
+		reg = (FW_IOCTL_LOG_DEQUEUE_IDX << FW_IOCTL_TYPE_SHIFT);
+		reg |= (log_index & 0xffff);
+		if (tegra->soc->has_bar2)
+			bar2_writel(tegra, reg, XUSB_BAR2_ARU_FW_SCRATCH);
+		else
+			fpci_writel(tegra, reg, XUSB_CFG_ARU_FW_SCRATCH);
+
+		dev_vdbg(dev, "new 0x%p log_index 0x%x\n", deq, (u32)log_index);
+	} else {
+		/* update dequeue pointer to firmware */
+		reg = (FW_IOCTL_LOG_DEQUEUE_LOW << FW_IOCTL_TYPE_SHIFT);
+		reg |= (physical_addr & 0xffff); /* lower 16-bits */
+		if (tegra->soc->has_bar2)
+			bar2_writel(tegra, reg, XUSB_BAR2_ARU_FW_SCRATCH);
+		else
+			fpci_writel(tegra, reg, XUSB_CFG_ARU_FW_SCRATCH);
+
+		reg = (FW_IOCTL_LOG_DEQUEUE_HIGH << FW_IOCTL_TYPE_SHIFT);
+		reg |= ((physical_addr >> 16) & 0xffff); /* higher 16-bits */
+		if (tegra->soc->has_bar2)
+			bar2_writel(tegra, reg, XUSB_BAR2_ARU_FW_SCRATCH);
+		else
+			fpci_writel(tegra, reg, XUSB_CFG_ARU_FW_SCRATCH);
+
+		dev_vdbg(dev, "new 0x%p physical addr 0x%x\n", deq, (u32)physical_addr);
+	}
+
+}
+
+static inline bool circ_buffer_full(struct circ_buf *circ)
+{
+	int space = CIRC_SPACE(circ->head, circ->tail, CIRC_BUF_SIZE);
+
+	return (space <= FW_LOG_SIZE);
+}
+
+static inline bool fw_log_available(struct tegra_xusb *tegra)
+{
+	return (tegra->log.dequeue->owner == DRIVER);
+}
+
+/**
+ * fw_log_wait_empty_timeout - wait firmware log thread to clean up shared
+ *      log buffer.
+ * @param tegra:        tegra_xusb context
+ * @param msec:         timeout value in millisecond
+ * @return true:        shared log buffer is empty,
+ *         false:       shared log buffer isn't empty.
+ */
+static inline bool fw_log_wait_empty_timeout(struct tegra_xusb *tegra,
+		unsigned int timeout)
+{
+	unsigned long target = jiffies + msecs_to_jiffies(timeout);
+	struct circ_buf *circ = &tegra->log.circ;
+	bool ret;
+
+	mutex_lock(&tegra->log.mutex);
+
+	while (fw_log_available(tegra) && time_is_after_jiffies(target)) {
+		if (circ_buffer_full(circ) &&
+			!test_bit(FW_LOG_FILE_OPENED, &tegra->log.flags))
+			break; /* buffer is full but nobody is reading log */
+
+		mutex_unlock(&tegra->log.mutex);
+		usleep_range(1000, 2000);
+		mutex_lock(&tegra->log.mutex);
+	}
+
+	ret = fw_log_available(tegra);
+	mutex_unlock(&tegra->log.mutex);
+
+	return ret;
+}
+
+/**
+ * fw_log_copy - copy firmware log from device's buffer to driver's circular
+ *      buffer.
+ * @param tegra tegra_xusb context
+ * @return true,        We still have firmware log in device's buffer to copy.
+ *                      This function returned due the driver's circular buffer
+ *                      is full. Caller should invoke this function again as
+ *                      soon as there is space in driver's circular buffer.
+ *         false,       Device's buffer is empty.
+ */
+static inline bool fw_log_copy(struct tegra_xusb *tegra)
+{
+	struct device *dev = tegra->dev;
+	struct circ_buf *circ = &tegra->log.circ;
+	int head, tail;
+	int buffer_len, copy_len;
+	struct log_entry *entry;
+	struct log_entry *first = tegra->log.virt_addr;
+
+	while (fw_log_available(tegra)) {
+
+		/* calculate maximum contiguous driver buffer length */
+		head = circ->head;
+		tail = READ_ONCE(circ->tail);
+		buffer_len = CIRC_SPACE_TO_END(head, tail, CIRC_BUF_SIZE);
+		/* round down to FW_LOG_SIZE */
+		buffer_len -= (buffer_len % FW_LOG_SIZE);
+		if (!buffer_len)
+			return true; /* log available but no space left */
+
+		/* calculate maximum contiguous log copy length */
+		entry = tegra->log.dequeue;
+		copy_len = 0;
+		do {
+			if (tegra->log.seq != entry->sequence_no) {
+				dev_warn(dev,
+					"%s: discontinuous seq no, expect %u get %u\n",
+					__func__, tegra->log.seq, entry->sequence_no);
+			}
+			tegra->log.seq = entry->sequence_no + 1;
+
+			copy_len += FW_LOG_SIZE;
+			buffer_len -= FW_LOG_SIZE;
+			if (!buffer_len)
+				break; /* no space left */
+			entry = fw_log_next(&tegra->log, entry);
+		} while ((entry->owner == DRIVER) && (entry != first));
+
+		memcpy(&circ->buf[head], tegra->log.dequeue, copy_len);
+		memset(tegra->log.dequeue, 0, copy_len);
+		circ->head = (circ->head + copy_len) & (CIRC_BUF_SIZE - 1);
+
+		mb(); /* make sure controller sees it */
+
+		fw_log_update_deq_pointer(&tegra->log, copy_len/FW_LOG_SIZE);
+
+		dev_vdbg(dev, "copied %d entries, new dequeue 0x%p\n",
+				copy_len/FW_LOG_SIZE, tegra->log.dequeue);
+		wake_up_interruptible(&tegra->log.read_wait);
+	}
+
+	return false;
+}
+
+static int fw_log_thread(void *data)
+{
+	struct tegra_xusb *tegra = data;
+	struct device *dev = tegra->dev;
+	struct circ_buf *circ = &tegra->log.circ;
+	bool logs_left;
+
+	dev_dbg(dev, "start firmware log thread\n");
+
+	do {
+		mutex_lock(&tegra->log.mutex);
+		if (circ_buffer_full(circ)) {
+			mutex_unlock(&tegra->log.mutex);
+			dev_info(dev, "%s: circ buffer full\n", __func__);
+			wait_event_interruptible(tegra->log.write_wait,
+				kthread_should_stop() ||
+				!circ_buffer_full(circ));
+			mutex_lock(&tegra->log.mutex);
+		}
+
+		logs_left = fw_log_copy(tegra);
+		mutex_unlock(&tegra->log.mutex);
+
+		/* relax if no logs left  */
+		if (!logs_left)
+			wait_event_interruptible_timeout(tegra->log.intr_wait,
+				fw_log_available(tegra), FW_LOG_THREAD_RELAX);
+	} while (!kthread_should_stop());
+
+	dev_dbg(dev, "stop firmware log thread\n");
+	return 0;
+}
+
+static inline bool circ_buffer_empty(struct circ_buf *circ)
+{
+	return (CIRC_CNT(circ->head, circ->tail, CIRC_BUF_SIZE) == 0);
+}
+
+static ssize_t fw_log_file_read(struct file *file, char __user *buf,
+		size_t count, loff_t *offp)
+{
+	struct tegra_xusb *tegra = file->private_data;
+	struct device *dev = tegra->dev;
+	struct circ_buf *circ = &tegra->log.circ;
+	int head, tail;
+	size_t n = 0;
+	int s;
+
+	mutex_lock(&tegra->log.mutex);
+
+	while (circ_buffer_empty(circ)) {
+		mutex_unlock(&tegra->log.mutex);
+		if (file->f_flags & O_NONBLOCK)
+			return -EAGAIN; /* non-blocking read */
+
+		dev_dbg(dev, "%s: nothing to read\n", __func__);
+
+		if (wait_event_interruptible(tegra->log.read_wait,
+				!circ_buffer_empty(circ)))
+			return -ERESTARTSYS;
+
+		if (mutex_lock_interruptible(&tegra->log.mutex))
+			return -ERESTARTSYS;
+	}
+
+	while (count > 0) {
+		head = READ_ONCE(circ->head);
+		tail = circ->tail;
+		s = min_t(int, count,
+				CIRC_CNT_TO_END(head, tail, CIRC_BUF_SIZE));
+
+		if (s > 0) {
+			if (copy_to_user(&buf[n], &circ->buf[tail], s)) {
+				dev_warn(dev, "copy_to_user failed\n");
+				mutex_unlock(&tegra->log.mutex);
+				return -EFAULT;
+			}
+			circ->tail = (circ->tail + s) & (CIRC_BUF_SIZE - 1);
+
+			count -= s;
+			n += s;
+		} else
+			break;
+	}
+
+	mutex_unlock(&tegra->log.mutex);
+
+	wake_up_interruptible(&tegra->log.write_wait);
+
+	dev_dbg(dev, "%s: %zu bytes\n", __func__, n);
+
+	return n;
+}
+
+static int fw_log_file_open(struct inode *inode, struct file *file)
+{
+	struct tegra_xusb *tegra;
+
+	file->private_data = inode->i_private;
+	tegra = file->private_data;
+
+	if (test_and_set_bit(FW_LOG_FILE_OPENED, &tegra->log.flags)) {
+		dev_info(tegra->dev, "%s: already opened\n", __func__);
+		return -EBUSY;
+	}
+
+	return 0;
+}
+
+static int fw_log_file_close(struct inode *inode, struct file *file)
+{
+	struct tegra_xusb *tegra = file->private_data;
+
+	clear_bit(FW_LOG_FILE_OPENED, &tegra->log.flags);
+
+	return 0;
+}
+
+static const struct file_operations firmware_log_fops = {
+	.open           = fw_log_file_open,
+	.release        = fw_log_file_close,
+	.read           = fw_log_file_read,
+	.owner          = THIS_MODULE,
+};
+
+static int fw_log_init(struct tegra_xusb *tegra)
+{
+	struct device *dev = tegra->dev;
+	int rc = 0;
+
+	if (!tegra->debugfs_dir)
+		return -ENODEV; /* no debugfs support */
+
+	if (test_bit(FW_LOG_CONTEXT_VALID, &tegra->log.flags))
+		return 0; /* already done */
+
+	/* allocate buffer to be shared between driver and firmware */
+	tegra->log.virt_addr = dma_alloc_coherent(dev,
+			FW_LOG_RING_SIZE, &tegra->log.phys_addr, GFP_KERNEL);
+
+	if (!tegra->log.virt_addr) {
+		dev_err(dev, "dma_alloc_coherent() size %d failed\n",
+				FW_LOG_RING_SIZE);
+		return -ENOMEM;
+	}
+
+	dev_info(dev, "%d bytes log buffer physical 0x%llx virtual 0x%p\n",
+			FW_LOG_RING_SIZE, tegra->log.phys_addr,
+			tegra->log.virt_addr);
+
+	memset(tegra->log.virt_addr, 0, FW_LOG_RING_SIZE);
+	tegra->log.dequeue = tegra->log.virt_addr;
+
+	tegra->log.circ.buf = vmalloc(CIRC_BUF_SIZE);
+	if (!tegra->log.circ.buf) {
+		rc = -ENOMEM;
+		goto error_free_dma;
+	}
+
+	tegra->log.circ.head = 0;
+	tegra->log.circ.tail = 0;
+
+	init_waitqueue_head(&tegra->log.read_wait);
+	init_waitqueue_head(&tegra->log.write_wait);
+	init_waitqueue_head(&tegra->log.intr_wait);
+
+	mutex_init(&tegra->log.mutex);
+
+	tegra->log.log_file = debugfs_create_file("firmware_log", 0444,
+			tegra->debugfs_dir, tegra, &firmware_log_fops);
+	if ((!tegra->log.log_file) ||
+			(tegra->log.log_file == ERR_PTR(-ENODEV))) {
+		dev_warn(dev, "debugfs_create_file() failed\n");
+		rc = -ENOMEM;
+		goto error_free_mem;
+	}
+
+	tegra->log.thread = kthread_run(fw_log_thread, tegra, "xusb-fw-log");
+	if (IS_ERR(tegra->log.thread)) {
+		dev_warn(dev, "kthread_run() failed\n");
+		rc = -ENOMEM;
+		goto error_remove_debugfs_file;
+	}
+
+	set_bit(FW_LOG_CONTEXT_VALID, &tegra->log.flags);
+	return rc;
+
+error_remove_debugfs_file:
+	debugfs_remove(tegra->log.log_file);
+error_free_mem:
+	vfree(tegra->log.circ.buf);
+error_free_dma:
+	dma_free_coherent(dev, FW_LOG_RING_SIZE,
+			tegra->log.virt_addr, tegra->log.phys_addr);
+	memset(&tegra->log, 0, sizeof(tegra->log));
+	return rc;
+}
+
+static void fw_log_deinit(struct tegra_xusb *tegra)
+{
+	struct device *dev = tegra->dev;
+
+	if (test_and_clear_bit(FW_LOG_CONTEXT_VALID, &tegra->log.flags)) {
+
+		debugfs_remove(tegra->log.log_file);
+
+		wake_up_interruptible(&tegra->log.read_wait);
+		wake_up_interruptible(&tegra->log.write_wait);
+		kthread_stop(tegra->log.thread);
+
+		mutex_lock(&tegra->log.mutex);
+		dma_free_coherent(dev, FW_LOG_RING_SIZE,
+				tegra->log.virt_addr, tegra->log.phys_addr);
+		vfree(tegra->log.circ.buf);
+		tegra->log.circ.head = tegra->log.circ.tail = 0;
+		mutex_unlock(&tegra->log.mutex);
+
+		mutex_destroy(&tegra->log.mutex);
+	}
+}
+
+static void tegra_xusb_debugfs_init(struct tegra_xusb *tegra)
+{
+	struct device *dev = tegra->dev;
+
+	tegra->debugfs_dir = debugfs_create_dir("tegra_xhci", NULL);
+	if (IS_ERR_OR_NULL(tegra->debugfs_dir)) {
+		dev_dbg(dev, "debugfs_create_dir() for tegra_xhci failed: %ld\n",
+			PTR_ERR(tegra->debugfs_dir));
+		tegra->debugfs_dir = NULL;
+		return;
+	}
+}
+
+static void tegra_xusb_debugfs_deinit(struct tegra_xusb *tegra)
+{
+	debugfs_remove(tegra->debugfs_dir);
+	tegra->debugfs_dir = NULL;
+}
+
 static int tegra_xusb_set_ss_clk(struct tegra_xusb *tegra,
 				 unsigned long rate)
 {
@@ -1011,7 +1499,6 @@ static int tegra_xusb_load_firmware_rom(struct tegra_xusb *tegra)
 	unsigned int code_tag_blocks, code_size_blocks, code_blocks;
 	struct tegra_xusb_fw_header *header;
 	struct device *dev = tegra->dev;
-	time64_t timestamp;
 	u64 address;
 	u32 value;
 	int err;
@@ -1024,6 +1511,16 @@ static int tegra_xusb_load_firmware_rom(struct tegra_xusb *tegra)
 		return 0;
 	}
 
+	if (header->build_log == LOG_MEMORY)
+		fw_log_init(tegra);
+
+	/* update the phys_log_buffer and total_entries here */
+	if (test_bit(FW_LOG_CONTEXT_VALID, &tegra->log.flags)) {
+		header->phys_addr_log_buffer =
+					cpu_to_le32(tegra->log.phys_addr);
+		header->total_log_entries = cpu_to_le32(FW_LOG_COUNT);
+	}
+
 	/* Program the size of DFI into ILOAD_ATTR. */
 	csb_writel(tegra, tegra->fw.size, XUSB_CSB_MP_ILOAD_ATTR);
 
@@ -1092,9 +1589,15 @@ static int tegra_xusb_load_firmware_rom(struct tegra_xusb *tegra)
 	if (tegra_xusb_wait_for_falcon(tegra))
 		return -EIO;
 
-	timestamp = le32_to_cpu(header->fwimg_created_time);
+	tegra->build_log = header->build_log;
+	tegra->version_id = header->version_id;
+	tegra->timestamp = le32_to_cpu(header->fwimg_created_time);
 
-	dev_info(dev, "Firmware timestamp: %ptTs UTC\n", &timestamp);
+	dev_info(dev, "Firmware timestamp: %ptTs UTC, Version: %2x.%02x %s\n",
+		&tegra->timestamp,
+		FW_MAJOR_VERSION(tegra->version_id),
+		FW_MINOR_VERSION(tegra->version_id),
+		(tegra->build_log == LOG_MEMORY) ? "debug" : "release");
 
 	return 0;
 }
@@ -1115,16 +1618,45 @@ static u32 tegra_xusb_read_firmware_header(struct tegra_xusb *tegra, u32 offset)
 
 static int tegra_xusb_init_ifr_firmware(struct tegra_xusb *tegra)
 {
-	time64_t timestamp;
+	u32 val;
 
 	if (tegra_xusb_wait_for_falcon(tegra))
 		return -EIO;
 
 #define offsetof_32(X, Y) ((u8)(offsetof(X, Y) / sizeof(__le32)))
-	timestamp = tegra_xusb_read_firmware_header(tegra, offsetof_32(struct tegra_xusb_fw_header,
-								       fwimg_created_time) << 2);
+	tegra->build_log = (tegra_xusb_read_firmware_header(tegra,
+				offsetof_32(struct tegra_xusb_fw_header,
+					num_hsic_port) << 2) >> 16) & 0xf;
+
+	if (tegra->build_log == LOG_MEMORY) {
+
+		fw_log_init(tegra);
 
-	dev_info(tegra->dev, "Firmware timestamp: %ptTs UTC\n", &timestamp);
+		/* set up fw log buffer address */
+		csb_writel(tegra, (u32) tegra->log.phys_addr,
+				XUSB_CSB_ARU_SCRATCH0);
+		csb_writel(tegra, (u32) (tegra->log.phys_addr >> 32),
+				XUSB_CSB_ARU_SCRATCH1);
+
+		/* set up fw log buffer size */
+		val = (FW_IOCTL_INIT_LOG_BUF << FW_IOCTL_TYPE_SHIFT);
+		val |= FW_LOG_COUNT;
+		bar2_writel(tegra, val, XUSB_BAR2_ARU_FW_SCRATCH);
+	}
+
+	tegra->timestamp = tegra_xusb_read_firmware_header(tegra,
+			offsetof_32(struct tegra_xusb_fw_header,
+				fwimg_created_time) << 2);
+
+	tegra->version_id = tegra_xusb_read_firmware_header(tegra,
+			offsetof_32(struct tegra_xusb_fw_header,
+				version_id) << 2);
+
+	dev_info(tegra->dev, "Firmware timestamp: %ptTs UTC, Version: %2x.%02x %s\n",
+		 &tegra->timestamp,
+		 FW_MAJOR_VERSION(tegra->version_id),
+		 FW_MINOR_VERSION(tegra->version_id),
+		 (tegra->build_log == LOG_MEMORY) ? "debug" : "release");
 
 	return 0;
 }
@@ -1902,6 +2434,8 @@ static int tegra_xusb_probe(struct platform_device *pdev)
 		goto disable_phy;
 	}
 
+	tegra_xusb_debugfs_init(tegra);
+
 	if (tegra->soc->firmware) {
 		err = tegra_xusb_request_firmware(tegra);
 		if (err < 0) {
@@ -2062,6 +2596,8 @@ static void tegra_xusb_remove(struct platform_device *pdev)
 	dma_free_coherent(&pdev->dev, tegra->fw.size, tegra->fw.virt,
 			  tegra->fw.phys);
 
+	fw_log_deinit(tegra);
+
 	if (tegra->padctl_irq)
 		pm_runtime_disable(&pdev->dev);
 
@@ -2069,6 +2605,8 @@ static void tegra_xusb_remove(struct platform_device *pdev)
 
 	tegra_xusb_disable(tegra);
 	tegra_xusb_padctl_put(tegra->padctl);
+
+	tegra_xusb_debugfs_deinit(tegra);
 }
 
 static void tegra_xusb_shutdown(struct platform_device *pdev)
@@ -2339,6 +2877,16 @@ static int tegra_xusb_enter_elpg(struct tegra_xusb *tegra, bool is_auto_resume)
 	if (wakeup)
 		tegra_xhci_enable_phy_sleepwalk_wake(tegra);
 
+	/* In ELPG, firmware log context is gone. Rewind shared log buffer. */
+	if (test_bit(FW_LOG_CONTEXT_VALID, &tegra->log.flags)) {
+		if (!circ_buffer_full(&tegra->log.circ)) {
+			if (fw_log_wait_empty_timeout(tegra, 500))
+				dev_info(tegra->dev, "%s still has logs\n", __func__);
+		}
+
+		tegra->log.dequeue = tegra->log.virt_addr;
+		tegra->log.seq = 0;
+	}
 	if (tegra->soc->has_pg_support)
 		tegra_xusb_powergate_partitions(tegra);
 	else {
-- 
2.43.0

